modules/raynRtdProvider.js

```diff
@@ -0,0 +1,198 @@
+ * @param {Array} segments
+ * @return {void}
+ */
+export function setSegmentsAsBidderOrtb2(bidConfig, bidders, integrationConfig, segments, checksum) {
+  const raynOrtb2 = {};
+
+  const raynContentData = [];
+  if (integrationConfig.iabContentCategories.v2_2.enabled && segments[checksum] && segments[checksum][6]) {
+    raynContentData.push(generateOrtbDataObject(6, segments[checksum][6], integrationConfig.iabContentCategories.v2_2.tier));
+  }
+  if (integrationConfig.iabContentCategories.v3_0.enabled && segments[checksum] && segments[checksum][7]) {
+    raynContentData.push(generateOrtbDataObject(7, segments[checksum][7], integrationConfig.iabContentCategories.v3_0.tier));
+  }
+  if (raynContentData.length > 0) {
+    deepSetValue(raynOrtb2, 'site.content.data', raynContentData);
+  }
+
+  if (integrationConfig.iabAudienceCategories.v1_1.enabled && segments[4]) {
+    const raynUserData = [generateOrtbDataObject(4, segments[4], integrationConfig.iabAudienceCategories.v1_1.tier)];
+    deepSetValue(raynOrtb2, 'user.data', raynUserData);
+  }
+
+  if (!bidders || bidders.length === 0 || !segments || Object.keys(segments).length <= 0) {
+    mergeDeep(bidConfig?.ortb2Fragments?.global, raynOrtb2);
+  } else {
+    const bidderConfig = Object.fromEntries(
+      bidders.map((bidder) => [bidder, raynOrtb2]),
+    );
+    mergeDeep(bidConfig?.ortb2Fragments?.bidder, bidderConfig);
+  }
+}
+
+/**
+ * Real-time data retrieval from Rayn
+ * @param {Object} reqBidsConfigObj
+ * @param {function} callback
+ * @param {Object} config
+ * @param {Object} userConsent
+ * @return {void}
+ */
+function alterBidRequests(reqBidsConfigObj, callback, config, userConsent) {
+  try {
+    const checksum = generateChecksum(window.location.href);
+
+    const segments = readSegments(RAYN_LOCAL_STORAGE_KEY);
+
+    const bidders = deepAccess(config, 'params.bidders');
+    const integrationConfig = mergeDeep(defaultIntegration, deepAccess(config, 'params.integration'));
+
+    if (segments && Object.keys(segments).length > 0 && (
+      segments[checksum] || (segments[4] &&
+        integrationConfig.iabAudienceCategories.v1_1.enabled &&
+        !integrationConfig.iabContentCategories.v2_2.enabled &&
+        !integrationConfig.iabContentCategories.v3_0.enabled
+      )
+    )) {
+      logMessage(LOG_PREFIX, `Segtax data from localStorage: ${JSON.stringify(segments)}`);
+      setSegmentsAsBidderOrtb2(reqBidsConfigObj, bidders, integrationConfig, segments, checksum);
+      callback();
+    } else if (window.raynJS && typeof window.raynJS.getSegtax === 'function') {
+      window.raynJS.getSegtax().then((segtaxData) => {
+        logMessage(LOG_PREFIX, `Segtax data from RaynJS: ${JSON.stringify(segtaxData)}`);
+        setSegmentsAsBidderOrtb2(reqBidsConfigObj, bidders, integrationConfig, segtaxData, checksum);
+        callback();
+      }).catch((error) => {
+        logError(LOG_PREFIX, error);
+        callback();
+      });
+    } else {
```

Comment:
why not fall back to communicating with your endpoint from the module?

Reply:
@patmmccann I'm not sure I understand what you mean exactly.
We have two steps, first we check for data in localStorage, then if our library is loaded we call it's function to retrieve the data and if none of the conditions are satisfied we just log a "no data" message and don't do anything.

Reply:
I see that. How about a third step; if neither local storage nor the library exist, you communicate with an endpoint to gather the segments?

Reply:
@patmmccann thank you for your observation. We may be adding this functionality at a later stage. For now we are relying on our customers to either run our library or populate local storage using our endpoints

Reply:
Please make sure your documentation reflects that